algorithms/CPlusPlus/Arrays/segregate0and1.cpp

```diff
@@ -0,0 +1,62 @@
+/*
+Description: A program to segregate the array of 0s and 1s
+
+Approach: Counting the number of 0s present in the array and 
+saving it in variable z. For the first z indexes, putting 0 and 
+for the remaining indexes putting 1.
+
+Time complexity: O(n)
+*/
+
+#include <iostream>
+#include <vector>
+using namespace std;
+
+void segregate0and1(vector<int> &v, int n)
+{
+
+  int z = 0;
+  for (int i = 0; i < n; i++)
+  {
+    //counting the number of 0s and storing it the variable z
+    if (v[i] == 0)
+    {
+      z++;
+    }
+  }
+  //for z indexes, putting 0 in the vector
+  //for remaining indexes putting 1 in the vector
+  for (int j = 0; j < n; j++)
+  {
+    if (j < z)
+    {
+      v[j] = 0;
+    }
+    else
+    {
+      v[j] = 1;
+    }
+  }
+}
+int main()
+{
+  int n;
+  cout << "Enter number of array elements\n";
+  cin >> n;
+  vector<int> v(n);
+  cout << "Enter the array elements (only 0 and 1)\n";
+
+  for (int i = 0; i < n; i++)
+  {
+    cin >> v[i];
+    v.push_back(v[i]);
```

Comment:
You are actually inserting the inputs twice to the vectors. Since you are using `n` and not `v.size()` in the loop your program works fine because your first `n` values will be the same as the input values you give.
```C
// If this is your inputs then you vector will look like the one below.
n -> 5
inputs: 1 0 1 0 1 

v.size() -> 10
vector -> {1 0 1 0 1 1 0 1 0 1}
```
```suggestion
    cin >> v[i];
```

Reply:
You should keep in mind that you can only use `cin >> v[i]` if you declare the vector with some size `n`, doing otherwise will cause a [segmentation fault](https://en.wikipedia.org/wiki/Segmentation_fault).

This is because when you declare it with size `n` ,during the compile time the memory of size `n*sizeof(data_type)` will be allocated for the vector and all the memory spaces will be initialised to `0` or a specified value, hence using `cin` or initialising it (eg:`v[1] = 5`) will be valid.

Doing such things without declaring the vector size will not be valid because no memory will be allocated for it to use, so initialising the vector values like `v[i] = i` will not be valid and causes `segfault`

In this case you can make use of `push_back()`. Whenever you try to `push_back()` , the vector will see if the present capacity is enough to accommodate that element, if it's enough the vector will assign the element in the next memory space, if not it will reallocate the memory and copy all the element from the present memory to the newly allocated memory.

Reply:
Thanks for the above description. I already made changes to the code.